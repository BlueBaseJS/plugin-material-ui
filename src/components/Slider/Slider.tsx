import { Theme, withStyles } from '@material-ui/core';

import FormControl from '@material-ui/core/FormControl';
import FormHelperText from '@material-ui/core/FormHelperText';
import FormLabel from '@material-ui/core/FormLabel';
import MUISlider from '@material-ui/lab/Slider';
import React from 'react';
import { SliderProps } from '@bluebase/components';
import { StyleSheet } from 'react-native';
import { componentMapper } from '@bluebase/component-mapper';

const styles = (theme: Theme): any => ({
	helperText: {
		marginTop: theme.spacing.unit * 2,
	},
	inlineSlider: {
		display: 'flex',
		flex: 1,
	},
	inlineValue: {
		marginLeft: theme.spacing.unit * 2,
	},
	label: {
		flex: 1,
		marginBottom: theme.spacing.unit * 2,
	},
	labelWrapper: {
		display: 'flex',
		flexDirection: 'row',
	},
	sliderWrapper: {
		alignItems: 'center',
		display: 'flex',
		flex: 1,
	},
	value: {
		marginBottom: theme.spacing.unit * 2,
	},
});

export const Slider = withStyles(styles)(
	componentMapper<SliderProps>(
		MUISlider,
		{
			max: 'maximumValue',
			min: 'minimumValue',

			onChange: ({ onChange, onValueChange }: any) => (event: any, value: number) => {
				if (onChange) {
					onChange(event, value);
				}

				if (onValueChange) {
					onValueChange(value);
				}
			},

			classes: 'classes',
			disabled: 'disabled',
			error: 'error',
			helperText: 'helperText',
			id: 'id',
			label: 'label',
			name: 'name',
			required: 'required',
			showValue: 'showValue',
			step: 'step',
			style: ({ style }: SliderProps) => StyleSheet.flatten(style),
			value: 'value',
		},
		{
			render(newProps: any, Component: any) {
				const { error, helperText, id, label, required, classes, showValue, ...props } = newProps;

				const formControlProps = {
					error,
					required,
				};

				// Label
				let labelNode: React.ReactNode = null;

				// Render a simple label
				if (label && !showValue) {
					labelNode = (
						<FormLabel className={classes.label} htmlFor={id}>
							{label}
						</FormLabel>
					);
				}
				// If we have label and show value, then we show value in line with the label
				else if (label && showValue) {
					labelNode = (
						<div className={classes.labelWrapper}>
							<FormLabel className={classes.label} htmlFor={id}>
								{label}
							</FormLabel>
							<FormLabel className={classes.value}>{props.value}</FormLabel>
						</div>
					);
				}

				let sliderNode = <Component id={id} {...props} />;

				// If theres no label then we show value in line with the slider
				if (!label && showValue) {
					sliderNode = (
						<div className={classes.sliderWrapper}>
							<div className={classes.inlineSlider}>{sliderNode}</div>
							<FormLabel className={classes.inlineValue}>{props.value}</FormLabel>
						</div>
					);
				}

				return (
					<FormControl {...formControlProps}>
						{labelNode}
						{sliderNode}
						{helperText ? (
							<FormHelperText className={classes.helperText}>{helperText}</FormHelperText>
						) : null}
					</FormControl>
				);
			},
		}
	)
) as React.ComponentType<SliderProps>;
