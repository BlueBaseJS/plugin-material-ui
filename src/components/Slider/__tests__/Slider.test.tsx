import React from 'react';
import { Slider } from '../Slider';
import { mount } from 'enzyme';

describe('Slider', () => {
	it('should map min and max props', () => {
		const component = mount(<Slider minimumValue={10} maximumValue={50} value={25} />);

		expect(component.find('Slider').first().prop('min')).toEqual(10);
		expect(component.find('Slider').first().prop('max')).toEqual(50);
		expect(component.find('Slider').first().prop('value')).toEqual(25);
	});

	it('should map onValueChange fn to onChange fn', () => {
		const cb = jest.fn();
		const component = mount(<Slider onValueChange={cb} />);

		const sw = component.find('Slider').first();
		const onChange = sw.prop('onChange') as any;

		onChange({ target: { value: sw.prop('value') } }, 32);

		// expect(component).toMatchSnapshot();
		expect(cb).toBeCalledTimes(1);
		expect(cb).toBeCalledWith(32);
	});

	it('should map onValueChange fn to onChange fn with value', () => {
		const cb = jest.fn();
		const component = mount(<Slider onValueChange={cb} />);

		const sw = component.find('Slider').first();
		const onChange = sw.prop('onChange') as any;

		onChange({ target: { value: sw.prop('value') } }, 43);

		// expect(component).toMatchSnapshot();
		expect(cb).toBeCalledTimes(1);
		expect(cb).toBeCalledWith(43);
	});

	it('should pass onChange as is if available', () => {
		const cb = jest.fn();

		const SLIDER = Slider as any;

		const component = mount(<SLIDER onChange={cb} />);

		const onChange = component.find('Slider').first().prop('onChange') as any;

		onChange('foo', 54);

		// expect(component).toMatchSnapshot();
		expect(cb).toBeCalledTimes(1);
		expect(cb).toBeCalledWith('foo', 54);
	});

	it('should display label & helperText', () => {
		const component = mount(<Slider label="FooBar" helperText="description" />);

		// expect(component).toMatchSnapshot();
		const label = component.find('FormLabel');
		const helperText = component.find('FormHelperText');

		expect(label.getElements()).toHaveLength(1);
		expect(label.text()).toBe('FooBar');

		expect(helperText.getElements()).toHaveLength(1);
		expect(helperText.text()).toBe('description');
	});

	it('should display label & value on top when showValue is true', () => {
		const component = mount(<Slider label="FooBar" showValue value={25} />);

		// expect(component).toMatchSnapshot();
		const label = component.find('[data-testid="value-on-top"] [data-testid="label"]').last();
		const value = component.find('[data-testid="value-on-top"] [data-testid="value"]').last();

		expect(label.text()).toBe('FooBar');
		expect(value.text()).toBe('25');
	});

	it('should display value on right when showValue is true and there is no label', () => {
		const component = mount(<Slider showValue value={25} />);

		// expect(component).toMatchSnapshot();
		const value = component.find('[data-testid="value-on-right"] [data-testid="value"]').last();

		expect(value.text()).toBe('25');
	});

	it('should NOT display label or helperText', () => {
		const component = mount(<Slider />);

		// expect(component).toMatchSnapshot();
		const label = component.find('FormLabel');
		const helperText = component.find('FormHelperText');

		expect(label.getElements()).toHaveLength(0);

		expect(helperText.getElements()).toHaveLength(0);
	});
});
