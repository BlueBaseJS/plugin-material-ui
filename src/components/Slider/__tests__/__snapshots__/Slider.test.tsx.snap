// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider should map min and max props 1`] = `
<WithStyles(Component)
  maximumValue={50}
  minimumValue={10}
  value={25}
>
  <Component
    classes={
      Object {
        "helperText": "Component-helperText-1",
        "inlineSlider": "Component-inlineSlider-2",
        "inlineValue": "Component-inlineValue-3",
        "label": "Component-label-4",
        "labelWrapper": "Component-labelWrapper-5",
        "sliderWrapper": "Component-sliderWrapper-6",
        "value": "Component-value-7",
      }
    }
    maximumValue={50}
    minimumValue={10}
    value={25}
  >
    <WithStyles(ForwardRef(FormControl))>
      <ForwardRef(FormControl)
        classes={
          Object {
            "fullWidth": "MuiFormControl-fullWidth",
            "marginDense": "MuiFormControl-marginDense",
            "marginNormal": "MuiFormControl-marginNormal",
            "root": "MuiFormControl-root",
          }
        }
      >
        <div
          className="MuiFormControl-root"
        >
          <Slider
            max={50}
            min={10}
            onChange={[Function]}
            value={25}
          >
            <ForwardRef(Slider)
              classes={
                Object {
                  "active": "MuiSlider-active",
                  "colorPrimary": "MuiSlider-colorPrimary",
                  "colorSecondary": "MuiSlider-colorSecondary",
                  "disabled": "Mui-disabled",
                  "focusVisible": "Mui-focusVisible",
                  "mark": "MuiSlider-mark",
                  "markActive": "MuiSlider-markActive",
                  "markLabel": "MuiSlider-markLabel",
                  "markLabelActive": "MuiSlider-markLabelActive",
                  "marked": "MuiSlider-marked",
                  "rail": "MuiSlider-rail",
                  "root": "MuiSlider-root",
                  "thumb": "MuiSlider-thumb",
                  "thumbColorPrimary": "MuiSlider-thumbColorPrimary",
                  "thumbColorSecondary": "MuiSlider-thumbColorSecondary",
                  "track": "MuiSlider-track",
                  "trackFalse": "MuiSlider-trackFalse",
                  "trackInverted": "MuiSlider-trackInverted",
                  "valueLabel": "MuiSlider-valueLabel",
                  "vertical": "MuiSlider-vertical",
                }
              }
              max={50}
              min={10}
              onChange={[Function]}
              value={25}
            >
              <span
                className="MuiSlider-root MuiSlider-colorPrimary"
                onMouseDown={[Function]}
              >
                <span
                  className="MuiSlider-rail"
                />
                <span
                  className="MuiSlider-track"
                  style={
                    Object {
                      "left": "0%",
                      "width": "37.5%",
                    }
                  }
                />
                <input
                  type="hidden"
                  value="25"
                />
                <WithStyles(ValueLabel)
                  className="MuiSlider-valueLabel"
                  disabled={false}
                  index={0}
                  key="0"
                  open={false}
                  value={25}
                  valueLabelDisplay="off"
                  valueLabelFormat={[Function]}
                >
                  <ValueLabel
                    className="MuiSlider-valueLabel"
                    classes={
                      Object {
                        "circle": "PrivateValueLabel-circle-35",
                        "label": "PrivateValueLabel-label-36",
                        "offset": "PrivateValueLabel-offset-34",
                        "open": "PrivateValueLabel-open-33",
                        "thumb": "PrivateValueLabel-thumb-32",
                      }
                    }
                    disabled={false}
                    index={0}
                    open={false}
                    value={25}
                    valueLabelDisplay="off"
                    valueLabelFormat={[Function]}
                  >
                    <span
                      aria-orientation="horizontal"
                      aria-valuemax={50}
                      aria-valuemin={10}
                      aria-valuenow={25}
                      className="MuiSlider-thumb MuiSlider-thumbColorPrimary"
                      data-index={0}
                      onBlur={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      onMouseLeave={[Function]}
                      onMouseOver={[Function]}
                      role="slider"
                      style={
                        Object {
                          "left": "37.5%",
                        }
                      }
                      tabIndex={0}
                    />
                  </ValueLabel>
                </WithStyles(ValueLabel)>
              </span>
            </ForwardRef(Slider)>
          </Slider>
        </div>
      </ForwardRef(FormControl)>
    </WithStyles(ForwardRef(FormControl))>
  </Component>
</WithStyles(Component)>
`;
